services/maven-metadata/maven-metadata.service.js

```diff
@@ -68,24 +80,71 @@ export default class MavenMetadata extends BaseXmlService {
     })
   }
 
-  async handle(_namedParams, { metadataUrl, versionPrefix, versionSuffix }) {
-    const data = await this.fetch({ metadataUrl })
-    let versions = data.metadata.versioning.versions.version.reverse()
-    if (versionPrefix !== undefined) {
-      versions = versions.filter(v => v.toString().startsWith(versionPrefix))
+  static applyFilter({ versions, filter }) {
+    if (!filter) {
+      return versions
     }
-    if (versionSuffix !== undefined) {
-      versions = versions.filter(v => v.toString().endsWith(versionSuffix))
+    return matcher(versions, filter)
+  }
+
+  static getLatestVersion({ data, strategy, filter }) {
+    if (strategy === 'latestProperty') {
+      if (data.metadata.versioning.latest === undefined) {
+        throw new InvalidResponse({
+          prettyMessage: "property 'latest' not found",
+        })
+      }
+      return data.metadata.versioning.latest
+    } else if (strategy === 'releaseProperty') {
+      if (data.metadata.versioning.release === undefined) {
+        throw new InvalidResponse({
+          prettyMessage: "property 'release' not found",
+        })
+      }
+      return data.metadata.versioning.release
+    } else if (strategy === 'highestVersion') {
+      if (
+        data.metadata.versioning.versions?.version === undefined ||
+        data.metadata.versioning.versions?.version.length === 0
```

Comment:
if versions.version is undefined,
`data.metadata.versioning.versions?.version` returns undefined
so `data.metadata.versioning.versions?.version.length` is same as `undefined.length`

```suggestion
        data.metadata.versioning.versions?.version?.length === 0
```